cli/klaus/src/listen.rs

```diff
@@ -0,0 +1,441 @@
+//! This module contains the functionality that listens for signals and process termination.
+//!
+//! On macOS this can be configured to listen for the termination of any process, but on Linux
+//! since we're using `prctl(2)` we are restricted to listening for the termination of the watchdog's
+//! parent process.
+//!
+//! On Linux `prctl(2)` is configured to deliver a user-specified signal (we chose SIGUSR1) when
+//! the target process terminates. This means that waiting for the notification is nonblocking on
+//! Linux. On macOS *waiting* for termination is blocking, but since we need to periodically check
+//! the shutdown flag, we use a poll-based method instead.
+//!
+//! Since we need to wait for a signal on Linux, we spawn a task that completes when a signal has
+//! been delivered. We also install a signal handler on macOS for feature parity. Depending on which
+//! signal is delivered we either trigger cleanup or simply shut down.
+//!
+//! On macOS since the termination notification is not delivered via signal, we need to spawn a task
+//! that will only complete once the notification has arrived. For compatibility between macOS and
+//! Linux implementations we also spawn a task on Linux, but this task is configured to never
+//! complete (otherwise the watchdog would exit before the signal is delivered). The task on macOS
+//! is created via `spawn_blocking` because there are no async calls in this task, but we still
+//! want the task to run in concert with the other async tasks (this is why we don't use a thread).
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Context;
+use futures::future::Either;
+use futures::{select, FutureExt, StreamExt};
+use nix::unistd::Pid;
+use signal_hook::consts::signal::*;
+use signal_hook_tokio::{Handle, Signals};
+use tokio::task::JoinHandle;
+use tracing::{debug, debug_span, error, info, info_span, Instrument};
+
+use crate::Error;
+
+const FLAG_CHECK_INTERVAL: Duration = Duration::from_millis(10);
+
+/// Produces a future that resolves when the shutdown flag is set.
+pub async fn wait_for_shutdown(flag: Arc<AtomicBool>) {
+    while !flag.load(std::sync::atomic::Ordering::SeqCst) {
+        tokio::time::sleep(FLAG_CHECK_INTERVAL).await;
+    }
+}
+
+/// Returns the PID that will be waited on
+pub fn target_pid(args: &crate::Cli) -> Pid {
+    #[cfg(target_os = "linux")]
+    let pid = if let Some(_pid) = args.pid {
+        debug!("ignoring user-provided PID, not available on Linux, using parent PID instead");
+        nix::unistd::getppid()
+    } else {
+        debug!("using parent PID for target PID");
+        nix::unistd::getppid()
+    };
+
+    #[cfg(target_os = "macos")]
+    let pid = if let Some(pid) = args.pid {
+        debug!("using user-provided target PID");
+        Pid::from_raw(pid)
+    } else {
+        debug!("using parent PID for target PID");
+        nix::unistd::getppid()
+    };
+
+    pid
+}
```

Comment:
This test hangs for me when run locally:
```
% just integ-tests services.bats -- -f generated
…
services.bats
   process-compose can run generated config file                                                                                                                                                            1/1 in 0 sec
```

It looks to be caused by a race condition where that non-interactive activation completes quicker than `klaus` takes to start up and `klaus` has been re-parented to PID1 by the time it calculates its parent PID:
```
(2) ~
% pstree -ws klaus
-+= 00001 root /sbin/launchd
 \--= 62255 dcarley /Users/dcarley/projects/flox/flox/cli/target/debug/klaus --logs /private/var/folders/8q/spckhr654cv4xrcv0fxsrlvc0000gn/T/nix-shell.lEkR69/bats-run-RPQ5hu/test/1/test/.flox/cache/klaus.62254.log --socket /private/tmp/flox.tests.7oduzR/tmp.82jjQyFQFs/run/flox.d220e316.sock --registry /var/folders/8q/spckhr654cv4xrcv0fxsrlvc0000gn/T/nix-shell.lEkR69/bats-run-RPQ5hu/test/1/flox-data/env-registry.json

(2) ~
% cat /private/var/folders/8q/spckhr654cv4xrcv0fxsrlvc0000gn/T/nix-shell.lEkR69/bats-run-RPQ5hu/test/1/test/.flox/cache/klaus.62254.log
2024-07-30T10:34:28.619439Z DEBUG klaus::sentry: No Sentry DSN set, skipping Sentry initialization
2024-07-30T10:34:28.619542Z DEBUG klaus: was provided a socket socket_path="/private/tmp/flox.tests.7oduzR/tmp.82jjQyFQFs/run/flox.d220e316.sock"
2024-07-30T10:34:28.619552Z DEBUG klaus: starting
2024-07-30T10:34:28.619645Z DEBUG klaus::listen: using parent PID for target PID
2024-07-30T10:34:28.619666Z  INFO klaus: watchdog is on duty this_pid=62255 target_pid=1
2024-07-30T10:34:28.619876Z DEBUG klaus::listen: task started task="macos_termination_listener"
2024-07-30T10:34:28.619896Z DEBUG klaus::listen: task started task="signal_listener"
```

In order to wait for `klaus` to start we'd have to use fiddly signals, IPC, or log inspection. Besides we don't really want to block activation on `klaus` starting as long as it can clean up after itself.

The other preferable way we could solve this is by always requiring a `--pid` argument, let `flox` calculate the parent when spawning, and consider it a bug to pass the PID of a process that isn't the parent.

It's kinda interesting that this was caught by a side-effect in BATS and seems a good reason not to close FD3 everywhere as I nearly did in https://github.com/flox/flox/pull/1824 but I can't think of a good way to test this more explicitly.

Reply:
Addressed in https://github.com/flox/flox/pull/1822/commits/263fcef9d2d784d59ddd22496cce434f25a4b6ee